### PR TITLE
For #21275 - Sort items by how many times they were actually shown

### PR DESCRIPTION
**Pull request by Mugurell**
Work based off of the not yet merged https://github.com/mozilla-mobile/fenix/pull/21104 which allows for the variations we need to ensure this code works.
Also needs https://github.com/mozilla-mobile/android-components/pull/10978.

Will mark this PR as draft until 21104 is merged.

To see the results: https://mozilla-hub.atlassian.net/browse/FNXV2-17682?focusedCommentId=458185

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by Mugurell** [comment]
Rebased on the current state of https://github.com/mozilla-mobile/fenix/pull/21104

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 1]
```diff
@@ -1127,6 +1139,91 @@ class HomeFragment : Fragment() {
...
+                                }
+                            )
+                        }
+                        false -> category
+                    }
+                }
+            }
+
+            homeFragmentStore.dispatch(HomeFragmentAction.PocketStoriesCategoriesChange(currentCategories))
+        }
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:1172]: I feel like there's too much happening here in the fragment:

* I want to bring up the idea of introducing a Middleware again: Ideally, I think, we should only dispatch an action here, saying "Those stories where shown to the user". The reducer updates the state accordingly. A Middleware may see this action and decide to persist this too.
* Even without the Middleware I think we should update the state in the reducer and not outside and just "injecting" a new state via an action.

Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:1158]: Updated.

**Review by pocmo — APPROVED** [review]

**Comment by Mugurell** [comment]
```
 SUITE: org.mozilla.fenix.browser.FenixSnackbarDelegateTest
    TEST: show with listener but no action
    FAILURE
 java.lang.InstantiationError: kotlin.jvm.functions.Function0
 ```
 Restarting CI.

**Pull request merged**